Apptainer definition file for PySR

wkharold opened the pull request:
fix: 686

This pull request adds an Apptainer definition file to PySR. Researchers can use it to build an Apptainer container that packages PySR and its dependencies (via a conda environment).

MilesCranmer commented:
Cool, thanks!

Since we have a `Dockerfile` already, could the Apptainer file be the same as that one, but in Apptainer syntax? I would prefer to avoid having two separate images. See https://github.com/MilesCranmer/PySR/blob/master/Dockerfile

Also, would be nice to have some continuous integration tests like we have for the docker file.

wkharold commented:
```Since we have a Dockerfile already, could the Apptainer file be the same as that one, but in Apptainer syntax?```

I had some trouble with that approach initially so I went with `conda`. Happily, what I learned doing the `conda` version allowed me to switch back to `pip3` and make it more like the Dockerfile.  I'll work on adding an integration test that mimics the Docker one.

MilesCranmer commented:
Thanks, could it also pull from the python image rather than rockylinux image? Since the base image is hosted by docker I don’t see a compelling reason for the base images to be different

MilesCranmer commented on Apptainer.def:
```diff
@@ -0,0 +1,58 @@
+
+	# Install PySR:
+	pip3 install --no-cache-dir .
+
```
```suggestion
        python -c "import pysr"
```
This triggers the installation of Julia libraries so should be done within the install step

wkharold replied:
Done. Note that because the container file system is read-only at runtime all the Julia stuff gets installed again, in the users $HOME/.pysr/ project, at runtime.

MilesCranmer replied:
> read-only at runtime all the Julia stuff gets installed again

Hm, I don't think this shouldn't happen. It should just use the read-only environment. Maybe one of the juliacall environment variables is not set to the apptainer dir? https://juliapy.github.io/PythonCall.jl/stable/juliacall/#julia-config

MilesCranmer replied:
Also see https://docs.julialang.org/en/v1/manual/environment-variables/#JULIA_LOAD_PATH and https://docs.julialang.org/en/v1/manual/environment-variables/#JULIA_DEPOT_PATH

MilesCranmer replied:
I guess it's tricky because Julia will also have the precompilation cache which might change depending on the user's environment variables. So ideally that should get written to a writeable directory, while the packages and registry are frozen by the image.

wkharold replied:
I've played around with this some and while you can get Julia to refer to the read-only environment using JULIA_LOAD_PATH and JULIA_DEPOT_PATH it is unhappy because the file system is read only. I'll go back and get the exact error and location in the code. I've solved this in other Apptainer contexts by using a writable overlay but I was hoping to be able to put the PySR container in a Google Artifact Registry repository so that it only had to be build once.

wkharold replied:
I'm testing a JULIA_DEPOT_PATH solution that looks promising. I should have a new commit today or tomorrow.

wkharold replied:
Fixed. Thanks for the JULIA_LOAD_PATH and JULIA_DEPOT_PATH pointers.

MilesCranmer commented on Apptainer.def:
```diff
@@ -0,0 +1,58 @@
+	# Download the version of Julia specified by the JULIA_RELEASE and JULIA_RELEASE_VERSION arguments
+	JULIA=julia-{{ JULIA_RELEASE }}.{{ JULIA_RELEASE_VERSION }}
+
+	wget https://julialang-s3.julialang.org/bin/linux/x64/{{ JULIA_RELEASE }}/${JULIA}-linux-x86_64.tar.gz
```
Will be done automatically by importing PySR. Alternatively we could pull from both the Julia docker image and the Python image, similar to the dockerfile

MilesCranmer commented on Apptainer.def:
```diff
@@ -0,0 +1,58 @@
+	wget https://julialang-s3.julialang.org/bin/linux/x64/{{ JULIA_RELEASE }}/${JULIA}-linux-x86_64.tar.gz
+
+	# Install it in /opt/julia
+	tar zxf ${JULIA}-linux-x86_64.tar.gz
```
Does apptainer work for arm64?

MilesCranmer replied:
Got it, I see. 

One other option is that we ditch the multi-stage builds, and just let the `import pysr` command install Julia via PythonCall.jl. It should be able to get the right binaries, and also simplifies the container.

wkharold commented:
Switched to bullseye for compatibility with the Dockerfile. I was using Rocky because that's what SchedMD uses as the base for the Slurm images built for Google's Cluster Toolkit (formerly HPC Toolkit). One my goals is to be able to bind mount the necessary Slurm commands and libs into a container so that you can use the Julia ClusterManagers interface to do distributed SR from within the container. Maybe your already know how to do that with Docker?

MilesCranmer commented:
> Switched to bullseye for compatibility with the Dockerfile. I was using Rocky because that's what SchedMD uses as the base for the Slurm images built for Google's Cluster Toolkit (formerly HPC Toolkit). One my goals is to be able to bind mount the necessary Slurm commands and libs into a container so that you can use the Julia ClusterManagers interface to do distributed SR from within the container. Maybe your already know how to do that with Docker?

I see, thanks. This is a good use-case. Since I have no particular need to use bullseye, so maybe we could switch the defaults of both the Dockerfile and the Apptainer to use Rocky? You can change the default here: https://github.com/MilesCranmer/PySR/blob/3aee19e38ceb3e0e1617d357a831400e01204658/Dockerfile#L6

MilesCranmer commented:
> One my goals is to be able to bind mount the necessary Slurm commands and libs into a container so that you can use the Julia ClusterManagers interface to do distributed SR from within the container. 

I do Slurm stuff with PySR, but haven't done it from a container before. In principle its possible though.

One other thing that is needed is a function to install *all* packages, even the Julia extensions like for ClusterManagers.jl. I just made a PR https://github.com/MilesCranmer/PySR/pull/688 to add this function. So if your apptainer file calls `pysr.load_all_packages()`, all the Julia packages ever needed by PySR (including ClusterManagers.jl) will be installed.

MilesCranmer commented:
Ok I merged https://github.com/MilesCranmer/PySR/pull/688 which will help for your stuff, do you want to merge those changes into your branch?

wkharold commented:
> I see, thanks. This is a good use-case. Since I have no particular need to use bullseye, so maybe we could switch the defaults of both the Dockerfile and the Apptainer to use Rocky? You can change the default here:

It doesn't look like there are Rocky based containers for Julia on Docker Hub. I think bullseye is OK for multicore work. Transitioning to Rocky would require significant changes and testing on Slurm to ensure that it worked properly.

MilesCranmer commented:
Oh sorry I thought you meant you wanted to switch to Rocky. I'm happy with Bullseye too though. As long as both of the container types can have overall similar images.

MilesCranmer reviewed (commented):
For adding unit-testing, I would edit this file:

https://github.com/MilesCranmer/PySR/blob/d7e87b4a825cae2190f71319fd21534758e7647f/.github/workflows/CI_docker.yml#L24-L34

and modify the build matrix to:

```yml
    strategy:
      matrix:
        os: [ubuntu-latest]
        arch: ['linux/amd64']
        container:
            - docker
            - apptainer
```

Then you can use

```yml
      if: ${{ matrix.container == "docker" }}
```

to run any docker-only steps, and similarly for apptainer. (I'm assuming there is a way to install apptainer in github actions.)

wkharold commented:
This is my first foray into Github CI testing. Would you prefer a modified CI_docker.yml file or a new CI_apptainer.yml file. My naive though is that the latter would be simpler but I could easily be missing something.

wkharold commented:
Scanning the test failures it looks like most/all of them are the result of the tests being run in a read-only directory. To avoid a `version module not found` error I cd into /pysr (which is in the container's read-only file system)  before doing anything involving `pysr`. Is there a python environment variable I could set to avoid having to do that?

MilesCranmer commented:
Is there a way you can mount the directory it uses for testing to a writeable directory?

MilesCranmer commented on .github/workflows/CI_apptainer.yml:
```diff
@@ -0,0 +1,37 @@
+      - name: Build apptainer
+        run: apptainer build pysr.sif Apptainer.def
+      - name: Test apptainer
+        run: apptainer exec ./pysr.sif /bin/bash -c 'pip install pytest nbval && PYTHONPATH=/pysr python3 -m pysr test main,cli,startup'
```
docker has the `-v` flag for mounting a directory. Does `apptainer exec` have something like that, so that the tests have some way to write files to? It would just be for the tests (since some of the unittests write and load files)

MilesCranmer replied:
Also wouldn't the `pip install pytests nbval` also be writing to the image here?

wkharold replied:
It looks like adding PYTHONPATH=/pysr obviates the need to switch to the /pysr directory. If the directory the tests are being run from is writable adding an additional mount point should be unnecessary.

MilesCranmer commented on .github/workflows/CI_apptainer.yml:
```diff
@@ -0,0 +1,37 @@
+        arch: ['linux/amd64']
+
+    steps:
+      - uses: actions/checkout@v2
```
```suggestion
      - uses: actions/checkout@v4
```

MilesCranmer commented on .github/workflows/CI_apptainer.yml:
```diff
@@ -0,0 +1,37 @@
+    strategy:
+      matrix:
+        os: [ubuntu-latest]
+        arch: ['linux/amd64']
```
```suggestion
```

MilesCranmer commented on Apptainer.def:
```diff
@@ -0,0 +1,60 @@
+	# Put the Julia dependencies in /pysr/depo
+	mkdir /pysr/depot
+
+	export JULIA_DEPOT_PATH="/pysr/depot:$JULIA_DEPOT_PATH"
```
```suggestion
	export JULIA_DEPOT_PATH="/pysr/depot"
```
(Or is there a reason to include `JULIA_DEPOT_PATH` here? I would have expected it to be empty)

wkharold replied:
I was just mimicking what they do in the Julia [docs](https://docs.julialang.org/en/v1/manual/environment-variables/#JULIA_DEPOT_PATH) they explain why there.

MilesCranmer replied:
I think those docs are written with a user environment in mind but since this is inside a container we likely want to strip references to external files?

wkharold replied:
I just tested without the trailing $JULIA_DEPOT_PATH. Seems to work fine. I'll take it out.

MilesCranmer commented on Apptainer.def:
```diff
@@ -0,0 +1,60 @@
+	pip3 install --no-cache-dir .
+
+	# Put the Julia dependencies in /pysr/depo
+	mkdir /pysr/depot
```
```suggestion
	mkdir /pysr/depot
	mkdir /pysr/env
```

MilesCranmer commented on Apptainer.def:
```diff
@@ -0,0 +1,60 @@
+	mkdir /pysr/depot
+
+	export JULIA_DEPOT_PATH="/pysr/depot:$JULIA_DEPOT_PATH"
+	export PYTHON_JULIAPKG_PROJECT="/pysr/depot"
```
```suggestion
	export PYTHON_JULIAPKG_PROJECT="/pysr/env"
```

MilesCranmer commented:
I tried this on my institutional cluster to see if I could use it locally, but it failed with:

```bash
> apptainer build pysr.sif Apptainer.def
INFO:    User not listed in /etc/subuid, trying root-mapped namespace
INFO:    fakeroot command not found
INFO:    Installing some packages may fail
FATAL:   Unable to build from Apptainer.def: while parsing definition: Apptainer.def: invalid section(s) specified: arguments
```

Is the `arguments` section new on 1.3.0` maybe? My institute looks to have 1.1.9.

wkharold commented:
>Is the arguments section new on 1.3.0` maybe? My institute looks to have 1.1.9. Maybe we could have compatibility for 1.1.0 (or even 1.0.0) to have broader support?

It got introduced in 1.2 as part of the `--build-arg / --build-arg-file` feature.  I don't have a problem with hardcoding the Julia/Python/distro bits to make it work with 1.x.

Lots of improvements have gone in since 1.1. You should encourage your institute sys admins to upgrade.

wkharold commented:
I thought adding PYTHONPATH=/psyr had fixed the 
```
ModuleNotFoundError: No module named 'pysr.version'
```
error. I no longer see it in my local builds or tests.

MilesCranmer commented:
> Lots of improvements have gone in since 1.1. You should encourage your institute sys admins to upgrade.

Good point. I asked them and they are planning on it in the few month timescale. I guess there will always be some clusters using older versions, so we might as well make it compatible with older versions.

MilesCranmer commented:
The error about `version.py` missing indicates that it might not be getting installed to the directory we expect. Or maybe when we import pysr, it is importing from the local directory, rather than `site-packages`. Basically, `version.py` only gets created when one runs `python setup.py install`.

wkharold commented:
I looked in the container and `version.py` ends up in `/pysr/pysr`. I added that to `PYTHONPATH` but the test still fails. I've copied the command from the log and pasted it into my shell and it works. There must be something different about the github environment.

MilesCranmer commented:
The GitHub actions logs show:

```
Installing collected packages: fastjsonschema, tornado, rpds-py, pyzmq, psutil, pluggy, platformdirs, nest-asyncio, iniconfig, debugpy, coverage, comm, attrs, referencing, pytest, jupyter-core, jupyter-client, jsonschema-specifications, jsonschema, ipykernel, nbformat, nbval
  WARNING: The script debugpy is installed in '/home/runner/.local/bin' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  WARNING: The scripts coverage, coverage-3.12 and coverage3 are installed in '/home/runner/.local/bin' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  WARNING: The scripts py.test and pytest are installed in '/home/runner/.local/bin' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  WARNING: The scripts jupyter, jupyter-migrate and jupyter-troubleshoot are installed in '/home/runner/.local/bin' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  WARNING: The scripts jupyter-kernel, jupyter-kernelspec and jupyter-run are installed in '/home/runner/.local/bin' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  WARNING: The script jsonschema is installed in '/home/runner/.local/bin' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  WARNING: The script jupyter-trust is installed in '/home/runner/.local/bin' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
```

which is odd, I thought it would install to paths within the image? Or are there other paths that need to be set before the pip installs?

wkharold commented:
Apptainer mounts the home directory of its initiator by default so those messages make sense but I didn't think about adding `/home/runner/.local/bin` to PATH. I'll add that and see if it changes things.

MilesCranmer commented:
Oh I see. But shouldn’t it be installed inside the container rather than the user’s home directory (which thus wouldn’t be saved in the container)?

MilesCranmer commented:
Maybe the Python `pip install` is trying to install to the home dir rather than an env?

MilesCranmer commented:
Looks like we can find this with `python -m site`: https://stackoverflow.com/a/46071447

wkharold commented:
I looked a little closer at my local environment and realized that I don't run the tests in the PySR directory. When I did that I reproduced the failure we're seeing, ugh. So I added steps to create a new directory specifically for testing in the CI action. I also experimented with the Apptainer `%test` section of the definition file and it seemed to simplify things so added that and changed `CI_apptainer.yml` to use it.

MilesCranmer reviewed (approved):
LGTM. Thanks for accommodating all of these suggested changes and in general for the great contribution!!
